File: singularity/implementation.js/Int32.js
/* Copyright 2021 ComdivByZero
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
(function() { 'use strict';

var module = {};
o7.export.Int32 = module;

var getjsa, assert;

var Size, Min, Max, min, max;

getjsa = o7.getjsa;
assert = o7.assert;

Size = 4;

Min = (1 << (Size * 8 - 2)) * -2;
Max = -1 - Min;

/* TODO */
min = o7.sarray(Size);
module.min = min;
max = o7.sarray(Size);
module.max = max;

function FromInt(v, i) {
	v = getjsa(v);
	v[0] = i & 0xFF;
	v[1] = (i >> 8) & 0xFF;
	v[2] = (i >> 16) & 0xFF;
	v[3] = (i >> 24) & 0xFF;
}
module.FromInt = FromInt;

function fromint(v, i) {
	assert(Min <= i && i <= Max);
	FromInt(v, i);
}

function toint(v) {
	v = getjsa(v);
	return v[0] | (v[1] << 8) | (v[2] << 16) | (v[3] << 24);
}

function ToInt(v) {
	var i;
	i = toint(v);
	assert(i != Min);
	return i;
}
module.ToInt = ToInt;

function Add(sum, a1, a2) {
	fromint(sum, toint(a1) + toint(a2));
}
module.Add = Add;

function Sub(diff, m, s) {
	fromint(diff, toint(m) - toint(s));
}
module.Sub = Sub;

function Mul(prod, m1, m2) {
	fromint(prod, toint(m1) * toint(m2));
}
module.Mul = Mul;

function Div(div, n, d) {
	fromint(div, toint(n) / toint(d) |0);
}
module.Div = Div;

function Mod(mod, n, d) {
	fromint(mod, toint(n) % toint(d));
}
module.Mod = Mod;

function DivMod(div, mod, n, d) {
	n = toint(n);
	d = toint(d);
	FromInt(div, n / d |0);
	FromInt(mod, n % d);
}
module.DivMod = DivMod;

function Cmp(l, r) {
	var c;
	l = toint(l);
	r = toint(r);
	if (l < r) {
		c = -1;
	} else if (l > r) {
		c = 1;
	} else {
		c = 0;
	}
	return c;
}
module.Cmp = Cmp;

function init() {
	fromint(min, Min);
	fromint(max, Max);
}

init();

return module;
})();

File: singularity/implementation.java/Int32.java
/* Copyright 2021-2022 ComdivByZero
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package o7;

import o7.O7;

public final class Int32 {

private static final int Size = 4, Min = (1 << (Size * 8 - 1)), Max = -1 + Min;

public static final byte[]
	min = new byte[Size],
	max = new byte[Size];

public static final int
	LittleEndian = 1,
	BigEndian    = 2,
	ByteOrder = LittleEndian;

public static void FromInt(byte[] v, int i) {
	v[0] = (byte)i;
	v[1] = (byte)(i >> 8);
	v[2] = (byte)(i >> 16);
	v[3] = (byte)(i >> 24);
}

public static void fromint(byte[] v, long i) {
	O7.asrt(Min <= i && i <= Max);
	FromInt(v, (int)i);
}

public static int toint(byte[] v) {
	return (0xFF & v[0]) | ((0xFF & v[1]) << 8) | ((0xFF & v[2]) << 16) | (v[3] << 24);
}

public static int ToInt(byte[] v) {
	int i;
	i = toint(v);
	O7.asrt(i > Min);
	return i;
}

public static void SwapOrder(byte[] v) {
	byte b;
	b = v[0]; v[0] = v[3]; v[3] = b;
	b = v[1]; v[1] = v[2]; v[2] = b;
}

public static void Add(byte[] sum, byte[] a1, byte[] a2) {
	fromint(sum, (long)toint(a1) + toint(a2));
}

public static void Sub(byte[] diff, byte[] m, byte[] s) {
	fromint(diff, (long)toint(m) - toint(s));
}

public static void Mul(byte[] prod, byte[] m1, byte[] m2) {
	fromint(prod, (long)toint(m1) * toint(m2));
}

public static void Div(byte[] div, byte[] n, byte[] d) {
	int ni, di;
	ni = toint(n);
	di = toint(d);
	O7.asrt(ni > Min || di != -1);
	FromInt(div, ni / di);
}

public static void Mod(byte[] mod, byte[] n, byte[] d) {
	int ni, di;
	ni = toint(n);
	di = toint(d);
	O7.asrt(ni > Min || di != -1);
	FromInt(mod, ni % di);
}

public static void DivMod(byte[] div, byte[] mod, byte[] n, byte[] d) {
	int ni, di;
	ni = toint(n);
	di = toint(d);
	O7.asrt(ni > Min || di != -1);
	FromInt(div, ni / di);
	FromInt(mod, ni % di);
}

public static int Cmp(byte[] l, byte[] r) {
	int li, ri, cmp;
	li = toint(l);
	ri = toint(r);
	if (li < ri) {
		cmp = -1;
	} else if (li != ri) {
		cmp = 1;
	} else {
		cmp = 0;
	}
	return cmp;
}

static {
	fromint(min, Min);
	fromint(max, Max);
}

}
